### ISSUE TITLE
Problems with MySQL.

### ISSUE THREAD

**Issue by 0xAF** [preaction/Yancy#13]
I have some troubles with MySQL tables.
Example collection:

	collections => {
		users => {
			properties => {
				username => {
					type => 'string',
				},
				password => {
					type => 'string',
				},
				email => {
					type => [ 'string', 'null' ],
					format => 'email',
				},
				phone => {
					type => 'string',
					#format => 'tel',
				},
				created => {
					type => 'string',
					format => 'date-time',
				},
		},
	}

Few problems arise.
1. If there is an empty `email` field, I get error: `OpenAPI >>> GET /yancy/api/users [{"message":"anyOf failed: Does not match email format.".....` in console and 500 in browser. The `email` is not required and `type` allows `null`. I'm not sure if this can be solved on Yancy side?
2. Not MySQL related, but as the [POD](https://metacpan.org/pod/distribution/Yancy/lib/Yancy/Help/Config.pod#Generated-Forms) suggests, there is `tel` format. It seems this is a mistake. I get `Format rule for 'tel' is missing at .../app/local/lib/perl5/JSON/Validator.pm line 615.` error on the console and the field is not rendered on the web.
3. The `created` column is `DATETIME` in MySQL. If I set the format to `date-time` in Yancy config, I get `OpenAPI >>> GET /yancy/api/users [{"message":"Does not match date-time format." .....` in console and 500 in browser. The [OpenAPI specification](https://swagger.io/docs/specification/data-models/data-types/) says: _the date-time notation as defined by RFC 3339, section 5.6, for example, 2017-07-21T17:32:28Z_. I cannot find a way to convert MySQL `DATETIME` to `RFC3339 DATE-TIME`. May be a `list_items` filter would be appropriate for such cases? So I can define filter to convert a DB field to something that OpenAPI will recognize later. As far as I understand, the filters now work only for the `create` and `delete` operations, not for `select`. I'm not sure if I'm missing something here or if this approach is correct.

Any suggestions are welcome. Thanks in advance.

**Comment by preaction** [comment]
Thanks for the feedback!

1. I'm going to have to look at this one more closely, but the likely place to solve this will be in JSON::Validator. To me, it sounds like it should work as-is: A `null` value cannot have a format and should not try validating the format. The [JSON Schema Validation RFC](http://json-schema.org/latest/json-schema-validation.html#rfc.section.7) doesn't fully explain what should happen in this situation, though, so we get to make it up I guess.

2. The warnings are expected: `tel` isn't defined as a format in the [JSON Schema Validation RFC](http://json-schema.org/latest/json-schema-validation.html#rfc.section.7), and so JSON::Validator doesn't know about it, but the field should still be rendering as an HTML5 `<input type="tel">`, so that's a bug. I should probably also find a way to silence those warnings from JSON::Validator.

3. The solution for the date/time format problem will be for the MySQL backend to translate the data in `DATETIME` fields to/from the required format. So, that's another bug to fix. Until that bug is fixed, it's likely best to use a `pattern => '<regex>'` to validate the date/time field in the MySQL format.

All of the backends are expected to do the same thing given the same schema, so translating date/time formats in the backend is the right thing. That date/time format problem does open up a bunch of possible problems with DBIx::Class and other backends, so I'll have to add some tests around those.

You're right in that filters are not executed after reading from the database: Filters are more about ensuring the data written to the database is correct. Filters were added primarily to solve password hashing for user authentication, but could also be used for processing data (processing Markdown into HTML, for example, or Base64 encoding an image once file upload fields are added to Yancy). I'm not fully happy with that whole system, and I suspect something better might replace it in the future (probably something using [Mojolicious::Validator](http://mojolicious.org/perldoc/Mojolicious/Validator#FILTERS)).

There's a case to be made about running filters after reading from the database, but there are also other ways to edit values coming from the database (editing the value in the template, or in a custom controller for example). If it becomes a common thing to need, then it might be a good feature for Yancy.

So, here's what needs to be done:

* [ ] Write a test for `type => [ 'string', 'null' ], format => '...'`, and fix it in JSON::Validator if necessary
* [ ] Figure out why `tel` fields are not displaying in the editor and fix it
* [ ] Find a way to silence warnings about formats that JSON::Validator does not recognize
* [ ] Write tests for date/time fields in the backends
* [ ] Translate RFC 3339 date/times to MySQL and back in the MySQL backend
 
I'll be able to get to this stuff in the coming week.

**Comment by preaction** [comment]
You can create virtual columns in the list view using [x-list-columns](https://metacpan.org/pod/distribution/Yancy/lib/Yancy/Help/Config.pod#x-list-columns): `x-list-columns => [ { title => 'Name', template => '{first_name} {last_name}' }]` ([standard caveats about first/last names apply](https://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/)), but there's presently no way to do that for the edit view.

I do want to add custom editor fields, which sounds like it could be used to do what you require. I could add something to the configuration like `x-field => { template => '<input />', value => '{first_name} {last_name}' }`. Then, as you mention, you could use `x-filter` to parse the field value into the values the database requires. That configuration can be part of the Yancy::Plugin::Form I'm developing to make sure that it's all using the same API. I'll add some notes about this in my development log.

But, overall, the Yancy editor isn't supposed to be a 100%, do-it-all solution. It provides something simple, direct, and usable for site admins. If the database can't be made to match what you want the data to look like, you'll need to write something custom for yourself (Yancy::Plugin::Form is about making it a lot easier to build custom forms, and in general the current thread of development is moving towards making it easy to build a fully custom Yancy editor out of reusable parts).

**Comment by 0xAF** [comment]
Thanks for the reply.
I should have read more carefully about the `x-list-columns`. Thanks for the pointer.
You mentioned the `pattern => { regex }` above. Can you please point me to where I can read more about it. I cannot find it in the JSON::Validator, OpenAPI (both in validator and mojo) and in Yancy. I've scanned the code of these modules, but I did not found anything to help me. I'm missing something.

As for the `tel`, it seems it's rendered, but I got confused when I was testing yesterday, because the web renders only the ID and the first field (Name or something else) and I had the impression that the rest fields are not rendered. Once you click `edit` or `new record` it's there. BTW, how do I choose which fields to be displayed in the table? Can't find my way either (no that I've searched deeply).

One more thing. If there is a db `table` (or `view` in my case) which has no `ID` field, I need to define it in the collection with `x-id-field`, so the `read_schema` works. Otherwise OpenAPI will fail and the app will not load. But as in my case with the `view` there is no column which can be used as unique id. So I tried to define `x-id-field => undef`, but this still do not work. Any ideas how this can be solved, or this will require my custom controller?

**Comment by preaction** [comment]
`pattern` is a string that works as a regex, like `pattern => '^\d{4}-\d{2}-\d{2}[T ]\d{2}:\d{2}:\d{2}$'`. I'll add some docs about it in Yancy::Help::Config.

`x-list-columns` is how you choose which data to displayed in the table/list of items.

Yancy presently can't work with tables that have no unique columns. There needs to be some way to uniquely identify a single row to edit, or else it might edit the wrong row(s). Yancy could be extended to support composite keys / multi-column keys, but it presently does not support that. Mostly I would suggest adding an auto-increment id field and just ignoring it in the things that don't care about it. There's a lot to be said for ensuring your database rows have a unique and meaningless ID field (using `email` as a primary key is just asking for trouble down the line).

There's also a possibility of adding support for tables with duplicate rows, but I suspect there might be problems there: The `UPDATE` statement would need to support things like `LIMIT` and `OFFSET` to make sure that the right row was targeted, or a toggle button to support "change one" / "change all" and the user would have to be made aware that which row is updated might not be the row they want updated. Still, certainly something that could be looked at if people wanted to do it. Likely not something that I'll do myself as a high priority though.

**Comment by preaction** [comment]
Oh, forgot to mention: The `date-time` problem seems to be that you've got a JSON::Validator version of 0.95 or less, which is weird because Yancy depends on Mojolicious::Plugin::OpenAPI version 1.22 which requires JSON::Validator > 0.95. I've added some tests for regression purposes, but they passed already. I'll release the new tests to CPAN Testers to see what they say (though most CPAN Testers can't test backends, since they need actual databases to test with).

**Comment by 0xAF** [comment]
I got the idea of the `pattern` at first. But I wondered what/where it is parsed, since I'm not quite familiar with the workflow of Yancy.

`x-list-columns`, got me for the second time. Thanks.

> Yancy presently can't work with tables that have no unique columns.

I've realized that. But still tried to undef the option. My next question was about multi-column keys, but you explained that already. In my case I can add the ID from one of the tables (from which the view is made)  into the view to help Yancy identify the rows. Anyway, this is view and it's not going to be modified, only displayed.

> There's also a possibility of adding support for tables with duplicate rows.

I guess this will take too much effort for quite small benefit. And it's a very custom case. But another option could be to use tables without unique keys only for display and disable changes on them, because I believe such tables are more likely to be views, not real tables. Or perhaps detect the views and disable the changing of the values on them. I'm not very familiar with other DBs, besides MySQL, SQLite and a bit of PG. But I believe it should be the same about the views everywhere, they cannot be changed.

Once I get some free time after 2-3 weeks, I can try to help with some of the minor stuff, like the multi-column key or some other stuff, until I get more familiar with Yancy. I like the idea of Yancy pretty much.


> The date-time problem seems to be that you've got a JSON::Validator version of 0.95 or less.

Hm, I'm using `carton` for this project and the cpanfile.snapshot shows that my JSON::Validator is version 2.05. I can confirm that from the Validator.pm's $VERSION string.

**Comment by preaction** [comment]
> > There's also a possibility of adding support for tables with duplicate rows.

> I guess this will take too much effort for quite small benefit. And it's a very custom case.

It's not that rare to have tables without primary keys. Many-to-many relationship tables generally don't have primary keys, since they consist of two foreign keys. They can usually be uniquely identified by the composite of the two columns though. But yeah, whether the benefit will outweigh the development and maintenance cost is up for debate.

> But another option could be to use tables without unique keys only for display and disable changes on them

Yancy could add a `x-readonly` configuration to the table which disables the save buttons, the add button, and removes the APIs for the same. And then, yeah, if Yancy is reading the schema and can't detect any unique keys, set the `x-readonly` flag on the table and log a warning or add a description with "Can't edit collection: No unique key detected." The user could then set their own description to remove that message.

I suspect part of the problem is that Yancy tries to display things that it has no way of editing, since it doesn't care about IDs until someone tries editing a thing. Yancy could also have some better validation of schemas to ensure that nothing is invalid (like, the id field(s) actually exist, or the collection is set to read-only).

> > The date-time problem seems to be that you've got a JSON::Validator version of 0.95 or less.

> Hm, I'm using carton for this project and the cpanfile.snapshot shows that my JSON::Validator is version 2.05. I can confirm that from the Validator.pm's $VERSION string.

Okay, that's super-weird. I added an explicit dependency anyway, since we do actually depend on JSON::Validator itself outside of Mojolicious::Plugin::OpenAPI, so maybe that'll fix it. The only way I've been able to reproduce the problem is by downgrading JSON::Validator, but it could also be that MySQL has configurable `DATETIME` formats and something would need to be added to Yancy::Backend::MySQL to support (ugh).

I've done all the stuff I mentioned doing above, and am releasing it in v1.003 now. Multi-column keys and `x-readonly` table configuration are both new features that can be added anytime, and PRs are most welcome :smiley:.
